src/document-click.js

```diff
@@ -248,35 +246,42 @@ export function onDocumentElementClick_(
   // Look for the referenced element.
   const hash = tgtLoc.hash.slice(1);
   let elem = null;
+
   if (hash) {
-    elem = ampdoc.getRootNode().getElementById(hash);
-    if (!elem) {
-      // Fallback to anchor[name] if element with id is not found.
-      // Linking to an anchor element with name is obsolete in html5.
-      elem = ampdoc.getRootNode().querySelector(`a[name=${hash}]`);
-    }
+    const escapedHash = escapeCssSelectorIdent(ampdoc.win, hash);
+    elem = (ampdoc.getRootNode().getElementById(hash) ||
+        // Fallback to anchor[name] if element with id is not found.
+        // Linking to an anchor element with name is obsolete in html5.
+        ampdoc.getRootNode().querySelector(`a[name=${escapedHash}]`));
   }
 
   // If possible do update the URL with the hash. As explained above
   // we do `replace` to avoid messing with the container's history.
   // The choice of `location.replace` vs `history.replaceState` is important.
   // Due to bugs, not every browser triggers `:target` pseudo-class when
   // `replaceState` is called. See http://www.zachleat.com/web/moving-target/
-  // for more details.
-  win.location.replace(`#${hash}`);
+  // for more details. Do this only if fragment has changed.
+  if (tgtLoc.hash != curLoc.hash) {
+    win.location.replace(`#${hash}`);
+  }
 
   // Scroll to the element if found.
   if (elem) {
+    // The first call to scrollIntoView overrides browsers' default
+    // scrolling behavior. The second call insides setTimeout allows us to
+    // scroll to that element properly.
+    // Without doing this, the viewport will not catch the updated scroll
+    // position on iOS Safari and hence calculate the wrong scrollTop for
+    // the scrollbar jumping the user back to the top for failing to calculate
+    // the new jumped offset.
+    // Without the first call there will be a visual jump due to browser scroll.
+    // See https://github.com/ampproject/amphtml/issues/5334 for more details.
     viewport./*OK*/scrollIntoView(elem);
+    setTimeout(() => viewport./*OK*/scrollIntoView(elem), 0);
   } else {
     dev().warn('documentElement',
         `failed to find element with id=${hash} or a[name=${hash}]`);
   }
-
-  // Push/pop history.
-  history.push(() => {
-    win.location.replace(`${curLoc.hash || '#'}`);
```

Comment:
Don't yet understand why this is being dropped. Can you re-sum this up in either thread? This exists so that users can click on back button and get to normal state, especially given the `:target` support.